core/server/common/src/main/java/alluxio/master/journal/JournalUtils.java

```diff
@@ -40,5 +52,100 @@ public static URI getJournalLocation() {
     }
   }
 
+  /**
+   * Writes a checkpoint of the entries in the given iterable.
+   *
+   * This is the complement of {@link #restoreJournalEntryCheckpoint(InputStream, Journaled)}.
+   *
+   * @param output the stream to write to
+   * @param iterable the iterable for fetching journal entries
+   */
+  public static void writeJournalEntryCheckpoint(OutputStream output, JournalEntryIterable iterable)
+      throws IOException, InterruptedException {
+    CheckpointOutputStream cos =
+        new CheckpointOutputStream(output, Constants.JOURNAL_ENTRY_CHECKPOINT_VERSION);
+    Iterator<JournalEntry> it = iterable.getJournalEntryIterator();
+    while (it.hasNext()) {
+      if (Thread.interrupted()) {
+        throw new InterruptedException();
+      }
+      it.next().writeDelimitedTo(cos);
+    }
+    cos.flush();
+  }
+
+  /**
+   * Restores the given journaled object from the journal entries in the input stream.
+   *
+   * This is the complement of
+   * {@link #writeJournalEntryCheckpoint(OutputStream, JournalEntryIterable)}.
+   *
+   * @param input the stream to read from
+   * @param journaled the object to restore
+   */
+  public static void restoreJournalEntryCheckpoint(InputStream input, Journaled journaled)
+      throws IOException {
+    CheckpointInputStream cis = new CheckpointInputStream(input);
+    if (cis.getVersion() != Constants.JOURNAL_ENTRY_CHECKPOINT_VERSION) {
```

Comment:
This is checkpoint version, not server version `JOURNAL_ENTRY_CHECKPOINT_VERSION` should never change, so newer server versions can read the checkpoints of older server versions. The purpose of this version is to differentiate between different types of checkpoints taken by the same object. E.g. the inode tree could be checkpointed by writing a bunch of journal entries *or* it could be checkpointed by serializing its data in another way. When we read the checkpoint, we can check the version and use it to decide how to deserialize the checkpoint.

Reply:
So, when would `cis.getVersion()` not equal `JOURNAL_ENTRY_CHECKPOINT_VERSION`?

If `JOURNAL_ENTRY_CHECKPOINT_VERSION` never changes, and we change inodeTree later, which which number changes?